vllm/core/block/token_ids.py

```diff
@@ -0,0 +1,244 @@
+from typing import Iterable, List, NamedTuple, Optional, Tuple, overload
+
+from vllm.sequence import Sequence
+
+
+class TokenRangeAnnotation(NamedTuple):
+    """
+    Annotates a range of placeholder tokens to capture content that will
+    replace them.
+    """
+
+    content_hash: int
```

Comment:
This might be a failure on my end, but where are these hashes actually computed? Are these hashes derived from the unprocessed multimodal data (i.e. raw image pixels for images)?

Will there need to be/is there already a way for the engine to automatically choose the appropriate hash function for a given modality?

Are all of these questions contingent on how #8346 gets integrated with this PR?

Reply:
Yup, they were -- originally #8346 took an approach of hiding more multi-modal logic away in each multi-modal model and I was going to do the same for hashing (i.e. delegate it to the model). But since I ended up propagating the placeholder ranges explicitly to the `Sequence` I updated this change to do the MM hashing there as well.